### ISSUE TITLE
Move TorchInductor Triton autotuner to compile time, use common cache

### ISSUE THREAD

**Issue by jansel** [pytorch/pytorch#93626]
We use the Triton autotuner here:
https://github.com/pytorch/torchdynamo/blob/8e4abe3f66fb6b8d97d6461d64f9c3be323a523e/torchinductor/triton_ops/autotune.py#L182
and, here:
https://github.com/pytorch/torchdynamo/blob/8e4abe3f66fb6b8d97d6461d64f9c3be323a523e/torchinductor/triton_ops/autotune.py#L216
To select tile sizes for generated kernels from a short list of options.  

This results in good performance, but has a high cost in terms of warm up time.  We will be running nearly identical autotuning runs to pick tilings that should be obvious.

I'd propose we create a small cost model to replace the autotuner.  As an input to this model we should take:
1) The size_hints (already rounded up to next power of two)
2) The indexing formulas of all the reads and writes
3) [Maybe] Number of `tmpX` variables used in kernel
4) [Maybe] Number of reduction loops in the kernel
5) [Maybe] Histogram of `ops.*` usage in the kernel

As an initial implementation, we could just build an on-disk cache/hash-table that runs the Triton autotuner on misses, but reuses a prior decision if the model inputs are identical.  This alone should dramatically speed up warmup times.

Longer term, we could take the data collected in this cache and train a small neural network to predict which tiling to use.  For this we should have the neural network predict the normalized runtime for each tiling choice.  This cost model can then replace the measurement step of the autotuner.

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @soumith @ngimel

**Comment by penguinwu** [comment]
@jansel Is this issue still relevant, or has it been done? We are trying to scrub old (before the 2.0 release) PT2 issues. Please suggest what to do next.

**Comment by jansel** [comment]
It is mostly done, though not completely.
